lambdas/data-migration2/tests/test-granulesAndFiles.js

```diff
@@ -591,6 +598,61 @@ test.serial('migrateGranuleRecord updates an already migrated record if the upda
   t.deepEqual(record.updated_at, new Date(newerGranule.updatedAt));
 });
 
+test.serial('migrateGranuleRecord supports undefined values in dynamo and overwrites defined values in Postgres when re-migrating', async (t) => {
+  const {
+    knex,
+    granulePgModel,
+    testCollection,
+    testExecution,
+  } = t.context;
+
+  const testGranule = generateTestGranule({
+    collectionId: constructCollectionId(testCollection.name, testCollection.version),
+    queryFields: { foo: cryptoRandomString({ length: 8 }) },
+    execution: testExecution.url,
+    status: 'completed',
+    updatedAt: Date.now() - 1000,
+  });
+
+  await createRejectableTransaction(knex, (trx) => migrateGranuleRecord(testGranule, trx));
+
+  const newerGranule = {
+    ...testGranule,
+    cmrLink: undefined,
+    queryFields: undefined,
+    updatedAt: Date.now(),
+  };
+
+  const granuleCumulusId = await createRejectableTransaction(
+    knex,
+    (trx) => migrateGranuleRecord(newerGranule, trx)
+  );
+  const record = await granulePgModel.get(knex, {
+    cumulus_id: granuleCumulusId,
+  });
+
+  // Convert the granule payload we're migrating to the Postgres Granule format
+  const expectedPgGranule = await translateApiGranuleToPostgresGranule({
+    dynamoRecord: newerGranule,
+    knexOrTransaction: knex,
+  });
+
+  t.teardown(async () => {
+    await t.context.granulePgModel.delete(t.context.knex, { cumulus_id: record.cumulus_id });
+  });
+
+  // The expectedPgGranule is an object created outside of PG and does not have foreign keys
+  // Remove null properties from the PG record before comparison. The second migration should have
+  // replaced valid values with null values.
+  const omitList = ['cumulus_id', 'pdr_cumulus_id', 'provider_cumulus_id'];
+  t.deepEqual(omit(removeNilProperties(record), omitList), expectedPgGranule);
+
+  // Redundant check explicitly asserting the undefined fields we passed in the second migration
+  // are null.
+  t.is(record.cmr_link, null);
+  t.is(record.query_fields, null);
+});
```

Comment:
I _think_ this is sufficient but if there are other tests/assertions here I'm not thinking of, I'm game.

Reply:
I think this test sufficiently covers the failure case, with the caveat that we're not addressing the granule.files issue *yet*, however we should err on the side of completeness and explicitly set all keys that can be to undefined or rather what a Dynamo lookup would return for a granule that only has the required keys in the API schema defined.